### PR TITLE
Add tests for var_value<Matrix> for univariate distributions

### PR DESCRIPTION
**Pull request by SteveBronder**
## Summary

This modifies the distribution test suite so that `var_value<Matrix>` is also tested.

## Tests

All the standard distribution tests but using `var_value<Matrix>`

## Side Effects

Nope

## Release notes

Tests distributions are `var_value<Matrix>` compatible

## Checklist

- [x] Math issue #1805 

- [x] Copyright holder: Steve Bronder

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

File: stan/math/prim/prob/poisson_binomial_lcdf.hpp [hunk 1]
```diff
@@ -50,8 +50,8 @@ return_type_t<T_theta> poisson_binomial_lcdf(const T_y& y,
   for (size_t i = 0; i < max_sz; ++i) {
     check_bounded(function, "Successes variable", y_vec[i], 0,
                   theta_vec[i].size());
-    check_finite(function, "Probability parameters", theta_vec[i]);
-    check_bounded(function, "Probability parameters", theta_vec[i], 0.0, 1.0);
+    check_finite(function, "Probability parameters", value_of(theta_vec[i]));
+    check_bounded(function, "Probability parameters", value_of(theta_vec[i]), 0.0, 1.0);
```
Comment (bbbales2) [inline stan/math/prim/prob/poisson_binomial_lcdf.hpp]: `theta_vec.val(i)`

File: stan/math/prim/prob/poisson_binomial_lpmf.hpp [hunk 2]
```diff
@@ -43,7 +43,7 @@ return_type_t<T_theta> poisson_binomial_lpmf(const T_y& y,
     check_bounded(function, "Successes variable", y_vec[i], 0,
                   theta_vec[i].size());
     check_finite(function, "Probability parameters", theta_vec[i]);
-    check_bounded(function, "Probability parameters", theta_vec[i], 0.0, 1.0);
+    check_bounded(function, "Probability parameters", value_of(theta_vec[i]), 0.0, 1.0);
```
Comment (bbbales2) [inline stan/math/prim/prob/poisson_binomial_lpmf.hpp]: `theta_vec.val(i)`

File: stan/math/prim/prob/bernoulli_logit_glm_lpmf.hpp [hunk 3]
```diff
@@ -80,7 +80,8 @@ return_type_t<T_x, T_alpha, T_beta> bernoulli_logit_glm_lpmf(
   }
 
   const auto& y_ref = to_ref(y);
-  check_bounded(function, "Vector of dependent variables", y_ref, 0, 1);
+  check_bounded(function, "Vector of dependent variables", value_of(y_ref), 0,
```
Comment (bbbales2) [inline stan/math/prim/prob/bernoulli_logit_glm_lpmf.hpp:83]: I think `y` will be integer here.

File: stan/math/prim/prob/bernoulli_logit_lpmf.hpp [hunk 4]
```diff
@@ -47,7 +47,7 @@ return_type_t<T_prob> bernoulli_logit_lpmf(const T_n& n, const T_prob& theta) {
   }
   T_n_ref n_ref = n;
   T_theta_ref theta_ref = theta;
-  check_bounded(function, "n", n_ref, 0, 1);
+  check_bounded(function, "n", value_of(n_ref), 0, 1);
```
Comment (bbbales2) [inline stan/math/prim/prob/bernoulli_logit_lpmf.hpp:50]: Integer, `value_of` not necessary

File: stan/math/prim/prob/bernoulli_lpmf.hpp [hunk 5]
```diff
@@ -41,8 +41,9 @@ return_type_t<T_prob> bernoulli_lpmf(const T_n& n, const T_prob& theta) {
                          "Probability parameter", theta);
   const T_n_ref n_ref = to_ref(n);
   const T_theta_ref theta_ref = to_ref(theta);
-  check_bounded(function, "n", n_ref, 0, 1);
-  check_bounded(function, "Probability parameter", theta_ref, 0.0, 1.0);
+  check_bounded(function, "n", value_of(n_ref), 0, 1);
```
Comment (bbbales2) [inline stan/math/prim/prob/bernoulli_lpmf.hpp:44]: Integer, value_of not necessary

File: stan/math/prim/prob/binomial_logit_lpmf.hpp [hunk 6]
```diff
@@ -61,7 +61,7 @@ return_type_t<T_prob> binomial_logit_lpmf(const T_n& n, const T_N& N,
   ref_type_t<decltype(value_of(N_arr))> N_val = value_of(N_arr);
   ref_type_t<decltype(value_of(alpha_arr))> alpha_val = value_of(alpha_arr);
 
-  check_bounded(function, "Successes variable", n_val, 0, N_val);
+  check_bounded(function, "Successes variable", value_of(n_val), 0, N_val);
```
Comment (bbbales2) [inline stan/math/prim/prob/binomial_logit_lpmf.hpp:64]: Integer

File: stan/math/prim/prob/binomial_lpmf.hpp [hunk 7]
```diff
@@ -51,9 +51,10 @@ return_type_t<T_prob> binomial_lpmf(const T_n& n, const T_N& N,
   T_N_ref N_ref = N;
   T_theta_ref theta_ref = theta;
 
-  check_bounded(function, "Successes variable", n_ref, 0, N_ref);
+  check_bounded(function, "Successes variable", value_of(n_ref), 0, N_ref);
```
Comment (bbbales2) [inline stan/math/prim/prob/binomial_lpmf.hpp:54]: Integer

File: stan/math/prim/prob/hypergeometric_lpmf.hpp [hunk 8]
```diff
@@ -18,7 +18,7 @@ template <bool propto, typename T_n, typename T_N, typename T_a, typename T_b>
 double hypergeometric_lpmf(const T_n& n, const T_N& N, const T_a& a,
                            const T_b& b) {
   static const char* function = "hypergeometric_lpmf";
-  check_bounded(function, "Successes variable", n, 0, a);
+  check_bounded(function, "Successes variable", value_of(n), 0, a);
```
Comment (bbbales2) [inline stan/math/prim/prob/hypergeometric_lpmf.hpp:21]: Integer

File: stan/math/prim/prob/hypergeometric_rng.hpp [hunk 9]
```diff
@@ -15,7 +15,7 @@ inline int hypergeometric_rng(int N, int a, int b, RNG& rng) {
   using boost::variate_generator;
   using boost::math::hypergeometric_distribution;
   static const char* function = "hypergeometric_rng";
-  check_bounded(function, "Draws parameter", N, 0, a + b);
+  check_bounded(function, "Draws parameter", value_of(N), 0, a + b);
```
Comment (bbbales2) [inline stan/math/prim/prob/hypergeometric_rng.hpp:18]: Integer

File: stan/math/prim/prob/ordered_logistic_glm_lpmf.hpp [hunk 10]
```diff
@@ -74,7 +74,8 @@ return_type_t<T_x, T_beta, T_cuts> ordered_logistic_glm_lpmf(
   T_cuts_ref cuts_ref = cuts;
   const auto& cuts_val = value_of_rec(cuts_ref);
   const auto& cuts_val_vec = to_ref(as_column_vector_or_scalar(cuts_val));
-  check_bounded(function, "Vector of dependent variables", y_ref, 1, N_classes);
+  check_bounded(function, "Vector of dependent variables", value_of(y_ref), 1,
```
Comment (bbbales2) [inline stan/math/prim/prob/ordered_logistic_glm_lpmf.hpp:77]: Integer

File: stan/math/prim/prob/ordered_logistic_lpmf.hpp [hunk 11]
```diff
@@ -103,7 +103,7 @@ return_type_t<T_loc, T_cut> ordered_logistic_lpmf(const T_y& y,
   const auto& lambda_arr = as_array_or_scalar(lambda_ref);
   ref_type_t<decltype(value_of(lambda_arr))> lambda_val = value_of(lambda_arr);
 
-  check_bounded(function, "Random variable", y_ref, 1, K);
+  check_bounded(function, "Random variable", value_of(y_ref), 1, K);
```
Comment (bbbales2) [inline stan/math/prim/prob/ordered_logistic_lpmf.hpp:106]: Integer

File: stan/math/prim/prob/poisson_binomial_lcdf.hpp [hunk 12]
```diff
@@ -50,8 +50,9 @@ return_type_t<T_theta> poisson_binomial_lcdf(const T_y& y,
   for (size_t i = 0; i < max_sz; ++i) {
     check_bounded(function, "Successes variable", y_vec[i], 0,
                   theta_vec[i].size());
-    check_finite(function, "Probability parameters", theta_vec[i]);
-    check_bounded(function, "Probability parameters", theta_vec[i], 0.0, 1.0);
+    check_finite(function, "Probability parameters", value_of(theta_vec[i]));
+    check_bounded(function, "Probability parameters", value_of(theta_vec[i]),
```
Comment (bbbales2) [inline stan/math/prim/prob/poisson_binomial_lcdf.hpp:54]: `theta_vec.val(i)`

File: stan/math/prim/prob/poisson_binomial_lpmf.hpp [hunk 13]
```diff
@@ -43,7 +43,8 @@ return_type_t<T_theta> poisson_binomial_lpmf(const T_y& y,
     check_bounded(function, "Successes variable", y_vec[i], 0,
                   theta_vec[i].size());
     check_finite(function, "Probability parameters", theta_vec[i]);
-    check_bounded(function, "Probability parameters", theta_vec[i], 0.0, 1.0);
+    check_bounded(function, "Probability parameters", value_of(theta_vec[i]),
```
Comment (bbbales2) [inline stan/math/prim/prob/poisson_binomial_lpmf.hpp:46]: `theta_vec.val(i)`

**Review by bbbales2 — CHANGES_REQUESTED** [review]
Took a skim. There were some value_ofs on stuff that I think are integers

File: stan/math/prim/prob/von_mises_lpdf.hpp [hunk 14]
```diff
@@ -86,9 +86,8 @@ return_type_t<T_y, T_loc, T_scale> von_mises_lpdf(T_y const& y, T_loc const& mu,
     }
   }
   if (!is_constant_all<T_scale>::value) {
-    const auto& bessel0 = modified_bessel_first_kind(0, kappa_val);
-    const auto& bessel1 = modified_bessel_first_kind(-1, kappa_val);
-    ops_partials.edge3_.partials_ = cos_mu_minus_y - bessel1 / bessel0;
+    ops_partials.edge3_.partials_ = cos_mu_minus_y -
+      modified_bessel_first_kind(-1, kappa_val) / modified_bessel_first_kind(0, kappa_val);
```
Comment (bbbales2) [inline stan/math/prim/prob/von_mises_lpdf.hpp:90]: If I did do this inline, I was getting memory problems with expressions.

**Comment by bbbales2** [comment]
Greencheck woooo. We got time to polish this a bit more since we aren't trying to merge varmats this release now.

File: stan/math/prim/prob/bernoulli_logit_lpmf.hpp (hunk 4 above)
Comment (SteveBronder) [reply stan/math/prim/prob/bernoulli_logit_lpmf.hpp:50]: Is it guranteed somewhere to always be an integer or do we only check that at the Stan language level?

Comment (bbbales2) [reply stan/math/prim/prob/bernoulli_logit_lpmf.hpp:50]: Stan language. These are the signatures:

```stan
real bernoulli_logit_lpmf(int, real)
real bernoulli_logit_lpmf(int, vector)
real bernoulli_logit_lpmf(int, row_vector)
real bernoulli_logit_lpmf(int, real[])
real bernoulli_logit_lpmf(int[], real)
real bernoulli_logit_lpmf(int[], vector)
real bernoulli_logit_lpmf(int[], row_vector)
real bernoulli_logit_lpmf(int[], real[])
```

File: stan/math/prim/prob/bernoulli_lpmf.hpp [hunk 15]
```diff
@@ -68,12 +69,12 @@ return_type_t<T_prob> bernoulli_lpmf(const T_n& n, const T_prob& theta) {
     if (sum == N) {
       logp += N * log(theta_dbl);
       if (!is_constant_all<T_prob>::value) {
-        ops_partials.edge1_.partials_[0] += N / theta_dbl;
+        ops_partials.edge1_.partials_[0] += N * inv(theta_dbl);
```
Comment (SteveBronder) [inline stan/math/prim/prob/bernoulli_lpmf.hpp:72]: What's up with this change?

Comment (SteveBronder) [reply stan/math/prim/prob/bernoulli_lpmf.hpp:72]: Oh this happened in a few places, why switch from `/` to `* inv()`?

Comment (bbbales2) [reply stan/math/prim/prob/bernoulli_lpmf.hpp:72]: https://github.com/stan-dev/math/issues/2307

File: test/prob/gamma/gamma_ccdf_log_test.hpp [hunk 16]
```diff
@@ -11,32 +11,26 @@ class AgradCcdfLogGamma : public AgradCcdfLogTest {
                     vector<double>& ccdf_log) {
     vector<double> param(3);
 
-    param[0] = 1.0;  // y
-    param[1] = 2.0;  // alpha
-    param[2] = 2.0;  // beta
+    param[0] = 0.7;  // y
+    param[1] = 1.1;  // alpha
+    param[2] = 0.9;  // beta
```
Comment (SteveBronder) [inline test/prob/gamma/gamma_ccdf_log_test.hpp:16]: Why did these change?

Comment (bbbales2) [reply test/prob/gamma/gamma_ccdf_log_test.hpp:16]: I changed these, but they can change back.

If you increase the difficulty of the tests by uncommenting the `.val_.d_` stuff [here](https://github.com/stan-dev/math/blob/034d00987d4106431711ba885edfc141710486b1/test/prob/utility.hpp#L97) you'll find lots of errors (things off by 1e-14 in high order derivatives) and if you fiddle with the values here you can get them to go away.

I added those `.val_.d_` tests with this pull request. They weren't there previously.

---

Short version cause that sounded confusing: the changes to the parameter values and the changes to `.val_.d_` intializers are both new and both can be reverted.

File: stan/math/prim/prob/bernoulli_logit_lpmf.hpp (hunk 4 above)
Comment (SteveBronder) [reply stan/math/prim/prob/bernoulli_logit_lpmf.hpp:50]: idk, idt there's anything bad about calling value_of here. Idt it would be an error and I don't think it would be a performance thing

File: stan/math/prim/prob/bernoulli_lpmf.hpp (hunk 15 above)
Comment (SteveBronder) [reply stan/math/prim/prob/bernoulli_lpmf.hpp:72]: I'm gonna add the static cast to double for the division here inf fvar's divide

**Comment by SteveBronder** [comment]
Alright I think with the division thing all that's left to do is sort out the integer value of and we are good. Personally idt it hurts and from the math library it is possible to have a `var` there

**Comment by bbbales2** [comment]
I cleaned out the test changes that didn't need to change and removed the value_ofs (gradients aren't going to work even if we do value_of if someone passes in a var). This pull is all good to me.

**Comment by SteveBronder** [comment]
All good to me as well! though idt I can approve because I'm the author

**Review by bbbales2 — APPROVED** [review]
